src/Job/NormalJob.ts

```diff
@@ -14,64 +15,110 @@ const DEFAULT_ENV: JobEnv = Object.freeze({
  * An implementation of {@link Job}.
  */
 export class NormalJob implements Job {
-	private prerequisites: Set<Job>
+	private prerequisiteJobs: Set<Job>
+
+	private prerequisiteFiles: Set<string>
+
+	private commands: string[]
+
+	private target: string
+
+	private environment: JobEnv
 
 	/**
-	 * @param name - The job's name. Must be unique between jobs in the same dependency graph.
-	 * @param prerequisites - A set containing all of this job's prerequisites. Defaults to no prerequisites.
-	 * @param environment - A description of the job's runtime environment. For forwards compatibility, always provide this parameter. Defaults to a Docker image containing GNU Make.
+	 * @param options - the settings for this Job.
+	 *
+	 * The default for options.prerequisiteJobs is the empty set.
+	 *
+	 * The default for options.prerequisiteFiles is the empty set.
+	 *
+	 * The default for options.environment is a Docker image containing GNU Make: {@link DEFAULT_ENV}. For forwards compatibility, always provide this parameter.
 	 */
-	constructor(
-		private readonly name: string,
-		prerequisites: Set<Job> = new Set(),
-		private environment: JobEnv = DEFAULT_ENV,
-	) {
-		// Make a copy so the caller can't directly access prerequisites. Encapsulation!
-		this.prerequisites = new Set(prerequisites)
+	constructor(options: NormalJobOptions) {
+		// Make copies so the caller can't directly access prerequisites. Encapsulation!
+		this.prerequisiteJobs = new Set(options.prerequisiteJobs) // We don't need a deep copy because Jobs are immutable.
+		this.prerequisiteFiles = new Set(options.prerequisiteFiles)
+		this.commands = options.commands.slice()
+		this.target = options.target
+		this.environment = options.environment
+			? {
+					dockerImage: options.environment.dockerImage,
+			  }
```

Comment:
You can automate this object cloning. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

Reply:
How would you do it?

Reply:
They have an example on that page under <q>Cloning an object</q>. Something like
```ts
this.environment = Object.assign({}, options.environment)
```
The benefit of this is that you don't need to know the properties of JobEnv

Reply:
`Object.assign` only shallow copies. Meaning if `JobEnv` gets a new property that's an object, someone could manipulate that and change a Job's internal state. I specified `dockerImage` explicitly so that any changes to `JobEnv` would have to be directly addressed here.

Reply:
Ah I see. If you want to get around that you can do a recursive object clone; there's plenty of examples of that out there.

> so that any changes to JobEnv would have to be directly addressed here.

I don't think this should be the case. `Job` isn't responsible for executing the job, only for tracking data about the job. So it's a bit unexpected if someone wants to change JobEnv and has to change it here as well. Relying on the internal details more than it needs to.